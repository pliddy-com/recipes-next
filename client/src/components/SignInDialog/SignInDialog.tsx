import { MouseEventHandler, SyntheticEvent, useState } from 'react';

import Button from '@mui/material/Button';
import Dialog from '@mui/material/Dialog';
import DialogTitle from '@mui/material/DialogTitle';
import DialogContent from '@mui/material/DialogContent';
import DialogContentText from '@mui/material/DialogContentText';
import DialogActions from '@mui/material/DialogActions';
import TextField from '@mui/material/TextField';

import Loading from 'components/Loading/Loading';

interface ISignInDialog {
  isLoading: boolean;
  isOpen: boolean;
  onClose: MouseEventHandler;
  onSignIn({
    email,
    password
  }: {
    email: string;
    password: string;
  }): Promise<void>;
}

const SignInDialog = ({
  isLoading,
  isOpen,
  onClose,
  onSignIn
}: ISignInDialog) => {
  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');
  const [emailError, setEmailError] = useState(false);
  const [passwordError, setPasswordError] = useState(false);

  const handleSubmit = async (event: SyntheticEvent) => {
    event.preventDefault();

    setEmailError(false);
    setPasswordError(false);

    if (email === '') {
      setEmailError(true);
    }
    if (password === '') {
      setPasswordError(true);
    }

    if (email && password) {
      try {
        await onSignIn({ email, password });
        setEmail('');
        setPassword('');
      } catch (e) {
        setEmailError(true);
        setPasswordError(true);
      }
    }
  };

  return (
    <Dialog
      open={isOpen}
      onClose={onClose}
      maxWidth="sm"
      className="signInDialog"
      data-testid="signInDialog"
    >
      <DialogTitle variant="h2" id="modal-title">
        Sign In
      </DialogTitle>
      <DialogContent>
        {isLoading ? (
          <Loading />
        ) : (
          <>
            <DialogContentText id="modal-description" variant="body2">
              Enter an email and password.
            </DialogContentText>
            <TextField
              autoFocus
              error={emailError}
              id="email"
              label="Email"
              onChange={(e) => setEmail(e.target.value)}
              margin="normal"
              size="small"
              type="email"
              value={email}
            />

            <TextField
              error={passwordError}
              id="password"
              label="Password"
              margin="normal"
              onChange={(e) => setPassword(e.target.value)}
              size="small"
              type="password"
              value={password}
            />
          </>
        )}
      </DialogContent>
      <DialogActions>
        <Button onClick={onClose} className="submit" aria-label="cancel">
          Cancel
        </Button>
        <Button
          aria-label="submit"
          className="submit"
          color="primary"
          type="submit"
          variant="contained"
          onClick={handleSubmit}
        >
          Submit
        </Button>
      </DialogActions>
    </Dialog>
  );
};

export default SignInDialog;
