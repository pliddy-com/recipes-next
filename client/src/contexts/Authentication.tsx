import {
  ReactElement,
  createContext,
  useContext,
  useEffect,
  useState
} from 'react';

import {
  AuthenticationDetails,
  CognitoUser,
  CognitoUserSession
} from 'amazon-cognito-identity-js';

import userPool from 'lib/userPool';

interface SignInProps {
  email: string;
  password: string;
}

interface AuthenticationContextValue {
  getToken(): Promise<null>;
  isAuth: boolean;
  isLoading: boolean;
  signIn({ email, password }: SignInProps): Promise<void>;
  signOut(): void;
}

const AuthenticationContext = createContext<AuthenticationContextValue>(
  {} as AuthenticationContextValue
);

interface AuthenticationProps {
  children: ReactElement | ReactElement[];
}

const AuthenticationProvider = (props: AuthenticationProps) => {
  const [isAuth, setIsAuth] = useState<boolean>(false);
  const [token, setToken] = useState<string | null>();
  const [isLoading, setIsLoading] = useState<boolean>(false);

  const getToken = async () => {
    const user = userPool && userPool.getCurrentUser();

    user &&
      user.getSession((err: Error | null, session: CognitoUserSession) => {
        if (err) {
          return null;
        }
        setToken(session.getIdToken().getJwtToken());
      });

    return null;
  };

  useEffect(() => {
    getToken();
    setIsAuth(token ? true : false);
  }, [token]);

  const signIn = async ({ email, password }: SignInProps) => {
    setIsLoading(true);
    await new Promise((resolve, reject) => {
      const user =
        userPool &&
        new CognitoUser({
          Username: email,
          Pool: userPool
        });

      const authDetails = new AuthenticationDetails({
        Username: email,
        Password: password
      });

      user &&
        user.authenticateUser(authDetails, {
          onSuccess: (result) => {
            setToken(result.getIdToken().getJwtToken());
            resolve(result);
            setIsLoading(false);
          },
          onFailure: (err) => {
            reject(err);
            setIsLoading(false);
          }
          // newPasswordRequired: (data) => {
          //   resolve(data);
          //   setIsLoading(false);
          // }
        });
    });
    setIsLoading(false);
  };

  const signOut = () => {
    const user = userPool && userPool.getCurrentUser();

    user && user.signOut();

    setToken(null);
  };

  return (
    <AuthenticationContext.Provider
      value={{
        getToken,
        isAuth,
        isLoading,
        signIn,
        signOut
      }}
    >
      {props.children}
    </AuthenticationContext.Provider>
  );
};

const useAuthContext = () => useContext(AuthenticationContext);

export { AuthenticationProvider, useAuthContext };
