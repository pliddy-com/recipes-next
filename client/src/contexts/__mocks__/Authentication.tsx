jest.createMockFromModule('../Authentication');

const authContextValues = {
  authLoading: false,
  isAuth: false,
  signIn: jest.fn(),
  signOut: jest.fn(),
  token: 'TOKEN'
};

const useAuthContext = jest.fn().mockImplementation(() => authContextValues);

export { useAuthContext };
