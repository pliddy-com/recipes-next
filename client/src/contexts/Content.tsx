/* istanbul ignore file */

import {
  Dispatch,
  ReactElement,
  SetStateAction,
  SyntheticEvent,
  createContext,
  useContext,
  useEffect,
  useState
} from 'react';

import { useRouter } from 'next/router';
import { useAuthContext } from './Authentication';
import { IRecipeChangeSet } from 'types/content';

interface ICMContext {
  canSave: boolean;
  editMode: boolean;
  editLoading: boolean;
  setCanSave: Dispatch<SetStateAction<boolean>>;
  setEditMode: Dispatch<SetStateAction<boolean>>;
  saveRecipe(event: SyntheticEvent): void;
  setRecipe: Dispatch<SetStateAction<IRecipeChangeSet | undefined>>;
  supressEdit: boolean;
  setSupressEdit: Dispatch<SetStateAction<boolean>>;
  toggleEdit(): void;
  uploadImage(file: File): void;
}

interface ContentManagementProps {
  children: ReactElement | ReactElement[];
}

const recipeApiUrl = `${process.env.NEXT_PUBLIC_AWS_API}/recipes`;
const assetsApiUrl = `${process.env.NEXT_PUBLIC_AWS_API}/assets`;

const ContentManagementContext = createContext<ICMContext>({} as ICMContext);

const ContentManagementProvider = (props: ContentManagementProps) => {
  const [editMode, setEditMode] = useState<boolean>(false);
  const [recipe, setRecipe] = useState<IRecipeChangeSet | undefined>();
  const [editLoading, setEditLoading] = useState<boolean>(false);
  const [canSave, setCanSave] = useState<boolean>(false);
  const [supressEdit, setSupressEdit] = useState<boolean>(false);

  const { isAuth, token } = useAuthContext();
  const { asPath, push } = useRouter();

  useEffect(() => {
    if (!isAuth) setEditMode(false);
  }, [isAuth]);

  const toggleEdit = () => {
    setEditMode(!editMode);
  };

  const updateEntry = async ({ recipe }: { recipe: IRecipeChangeSet }) => {
    if (!isAuth || !token) {
      throw new Error('User is not authenticated.');
    }

    if (!canSave) {
      throw new Error('Content has not changed.');
    }

    const restApi = `${recipeApiUrl}/${recipe.id}`;

    try {
      const response = await fetch(restApi, {
        method: 'PUT',
        headers: {
          Authorization: `${token}`
        },
        body: JSON.stringify(recipe)
      });

      return response.json();
    } catch (e) {
      console.error('Could not save changes to the recipe.');
    }
  };

  const saveRecipe = async (event: SyntheticEvent) => {
    setEditLoading(true);
    event.preventDefault();

    if (!isAuth || !token) {
      throw new Error('User is not authenticated');
    }

    if (recipe) {
      try {
        recipe && (await updateEntry({ recipe }));

        const page = asPath.split('/').slice(-1)[0];
        if (recipe && page !== recipe.slug) push('/');

        setEditMode(false);
        setEditLoading(false);
      } catch (e) {
        // TODO: handle error in UI
        console.error(e);
      }
    } else {
      const err = new Error('No recipe payload to save.');
      throw err;
    }
  };

  // upload media asset to Contentful

  // MOVE TO AWS API
  // UPLOAD IMAGE TO S3 USING SECURE URL
  // SEND SECURE DOWNLOAD URL TO CONTENTFUL WITH CREATE ASSET REQUEST

  const bufferToText = (buffer: ArrayBufferLike) => {
    const bufferByteLength = buffer.byteLength;
    const bufferUint8Array = new Uint8Array(buffer, 0, bufferByteLength);

    return new TextDecoder().decode(bufferUint8Array);
  };

  const uploadImage = async (file: File) => {
    try {
      if (!isAuth || !token) {
        throw new Error('User is not authenticated.');
      }

      console.log({ file });

      // const { name, type } = file;

      const buffer = await file.arrayBuffer();
      const body = bufferToText(buffer);

      console.log('body', body);

      try {
        const response = await fetch(assetsApiUrl, {
          method: 'POST',
          headers: {
            Authorization: `${token}`
          },
          body: JSON.stringify(body)
        });

        return response.json();
      } catch (e) {
        console.error('Could not upload the image.');
      }

      // const upload =
      //   buffer &&
      //   (await client
      //     .getSpace(SPACE_ID)
      //     .then((space) => space.getEnvironment('master'))
      //     .then((environment) =>
      //       environment.createAssetFromFiles({
      //         fields: {
      //           title: {
      //             'en-US': name
      //           },
      //           description: {
      //             'en-US': 'Asset description'
      //           },
      //           file: {
      //             'en-US': {
      //               contentType: type,
      //               fileName: name,
      //               file: buffer
      //             }
      //           }
      //         }
      //       })
      //     )
      //     .then((asset) => {
      //       console.log('processing upload');
      //       return asset.processForAllLocales({ processingCheckWait: 2000 });
      //     })
      //     .then((asset) => {
      //       console.log('publishing upload');
      //       return asset.publish();
      //     })
      //     .then((asset) => {
      //       console.log({ asset });
      //       return asset;
      //     }));

      // console.log('Upload ', upload);
    } catch (e) {
      console.error(e);
    }
  };

  return (
    <ContentManagementContext.Provider
      value={{
        canSave,
        editLoading,
        editMode,
        saveRecipe,
        setCanSave,
        setEditMode,
        setRecipe,
        setSupressEdit,
        supressEdit,
        toggleEdit,
        uploadImage
      }}
    >
      {props.children}
    </ContentManagementContext.Provider>
  );
};

const useContentManagementContext = () => useContext(ContentManagementContext);

export { ContentManagementProvider, useContentManagementContext };
