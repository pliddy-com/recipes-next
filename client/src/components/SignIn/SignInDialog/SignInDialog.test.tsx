import '@testing-library/jest-dom';
import { act, fireEvent, render, waitFor } from '@testing-library/react';

import SignInDialog from './SignInDialog';

import * as AuthContext from 'contexts/Authentication';

jest.mock('contexts/Authentication');

jest.mock('next/router', () => ({
  useRouter: jest.fn().mockReturnValue({
    push: jest.fn(),
    back: jest.fn(),
    events: {
      on: jest.fn(),
      off: jest.fn()
    }
  })
}));

describe('SignInDialog', () => {
  describe('when isAuth is false', () => {
    it('it renders a sign in form', async () => {
      const emailValue = 'test@mail.com';
      const passwordValue = 'password';

      const { asFragment, queryByTestId, queryByLabelText, queryByRole } =
        render(
          <SignInDialog
            isLoading={false}
            onSignIn={jest.fn()}
            isOpen={true}
            onClose={jest.fn()}
          />
        );

      // wait for dynamic component to load
      await act(async () =>
        waitFor(() => expect(queryByTestId('signInDialog')).toBeInTheDocument())
      );

      // expect(authSpy).toBeCalled();

      const submitButton = queryByRole('button', {
        name: 'submit'
      });

      // assert that form is validated on submit
      submitButton && fireEvent.click(submitButton);

      // Test form field inputs
      const email_input = queryByLabelText('Email');
      const password_input = queryByLabelText('Password');

      email_input &&
        fireEvent.change(email_input, {
          target: { value: emailValue }
        });

      password_input &&
        fireEvent.change(password_input, {
          target: { value: passwordValue }
        });

      expect(email_input).toHaveValue(emailValue);
      expect(password_input).toHaveValue(passwordValue);

      waitFor(() => submitButton && fireEvent.click(submitButton));

      // assert that the component matches the existing snapshot
      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('when signIn returns an error', () => {
    it('it handles the error', async () => {
      const emailValue = 'test@mail.com';
      const passwordValue = 'password';
      const testError = 'sign in failed';

      const contextValues = {
        editMode: false,
        getToken: jest.fn(),
        isAuth: false,
        isLoading: false,
        saveRecipe: jest.fn(),
        setRecipe: jest.fn(),
        signIn: jest.fn(),
        signOut: jest.fn(),
        toggleEdit: jest.fn()
      };

      jest
        .spyOn(AuthContext, 'useAuthContext')
        .mockImplementation(() => contextValues);

      const { queryByTestId, queryByLabelText, queryByRole } = render(
        <SignInDialog
          isLoading={false}
          onSignIn={jest.fn().mockRejectedValueOnce('error')}
          isOpen={true}
          onClose={jest.fn()}
        />
      );

      // wait for dynamic component to load
      await act(async () =>
        waitFor(() => expect(queryByTestId('signInDialog')).toBeInTheDocument())
      );

      // Test form field inputs
      const email_input = queryByLabelText('Email');
      const password_input = queryByLabelText('Password');

      email_input &&
        fireEvent.change(email_input, {
          target: { value: emailValue }
        });

      password_input &&
        fireEvent.change(password_input, {
          target: { value: passwordValue }
        });

      const submitButton = queryByRole('button', {
        name: 'submit'
      });

      if (submitButton) {
        try {
          await waitFor(() => fireEvent.click(submitButton));
        } catch (e) {
          console.error(e);
          expect(e).toEqual(new Error(testError));
        }
      }
    });
  });

  describe('when isLoading is true', () => {
    it('it renders a Loading indicator', async () => {
      const { asFragment, queryByTestId, queryByRole } = render(
        <SignInDialog
          isLoading={true}
          onSignIn={jest.fn()}
          isOpen={true}
          onClose={jest.fn()}
        />
      );

      // wait for dynamic component to load
      await act(async () =>
        waitFor(() => expect(queryByTestId('signInDialog')).toBeInTheDocument())
      );

      const component = queryByRole('graphics-symbol', {
        name: 'spinner'
      });
      expect(component).toBeInTheDocument();

      // assert that the component matches the existing snapshot
      expect(asFragment()).toMatchSnapshot();
    });
  });
});
