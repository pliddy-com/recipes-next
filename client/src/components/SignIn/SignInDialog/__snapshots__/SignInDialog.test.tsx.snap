// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignInDialog when isAuth is false it renders a sign in form 1`] = `<DocumentFragment />`;

exports[`SignInDialog when isLoading is true it renders a Loading indicator 1`] = `<DocumentFragment />`;
